File: src/effects/color_wipe.rs
use std::time::Duration;
use crate::effects::{Effect, Buffer};
use crate::pixels::Color;

enum WipeState {
    Wiping,
    Clearing,
}

pub struct ColorWipeEffect {
    color: Color,
    speed_ms: f32,
    state: WipeState,
    pos: usize,
    timer: f32,
}

impl ColorWipeEffect {
    pub fn new(color: Color, speed_ms: f32) -> Self {
        Self {
            color,
            speed_ms,
            state: WipeState::Wiping,
            pos: 0,
            timer: 0.0,
        }
    }
}

impl Effect for ColorWipeEffect {
    fn name(&self) -> &'static str { "Color Wipe" }

    fn update(&mut self, buffer: &mut Buffer, delta: Duration) -> bool {
        let n = buffer.len();
        self.timer += delta.as_secs_f32() * 1000.0;

        while self.timer >= self.speed_ms {
            self.timer -= self.speed_ms;
            match self.state {
                WipeState::Wiping => {
                    if self.pos < n {
                        buffer[self.pos] = self.color;
                        self.pos += 1;
                    }
                    if self.pos >= n {
                        self.pos = 0;
                        self.state = WipeState::Clearing;
                    }
                }
                WipeState::Clearing => {
                    if self.pos < n {
                        buffer[self.pos] = [0, 0, 0];
                        self.pos += 1;
                    }
                    if self.pos >= n {
                        self.pos = 0;
                        self.state = WipeState::Wiping;
                    }
                }
            }
        }

        true
    }
}

File: effects/color_wipe.lua
name = "Color Wipe"
description = "Fills the strip with the user-selected color one pixel at a time, then clears it the same way."

local speed_ms = 20.0
local state    = "wiping"
local pos      = 0
local timer    = 0.0

function init(n)
    pos   = 0
    state = "wiping"
    timer = 0.0
end

function update(buf, dt)
    local n = buf:len()
    local r = COLOR_R or 255
    local g = COLOR_G or 255
    local b = COLOR_B or 255
    timer = timer + dt * 1000.0

    while timer >= speed_ms do
        timer = timer - speed_ms
        if state == "wiping" then
            if pos < n then
                buf:set(pos, r, g, b)
                pos = pos + 1
            end
            if pos >= n then pos = 0; state = "clearing" end
        else
            if pos < n then
                buf:set(pos, 0, 0, 0)
                pos = pos + 1
            end
            if pos >= n then pos = 0; state = "wiping" end
        end
    end
    return true
end
